upb/bindings/lua/upb.c

```diff
@@ -1056,6 +1064,128 @@ static const struct luaL_Reg lupb_enumdef_m[] = {
+
+void lupb_filedef_pushwrapper(lua_State *L, const upb_filedef *f,
+                              const void *ref_donor) {
+  lupb_refcounted_pushwrapper(L, upb_filedef_upcast(f), LUPB_FILEDEF, ref_donor,
+                              sizeof(void *));
+}
+
+void lupb_filedef_pushnewrapper(lua_State *L, const upb_filedef *f,
+                                const void *ref_donor) {
+  lupb_refcounted_pushnewrapper(L, upb_filedef_upcast(f), LUPB_FILEDEF,
+                                ref_donor);
+}
+
+const upb_filedef *lupb_filedef_check(lua_State *L, int narg) {
+  return lupb_refcounted_check(L, narg, LUPB_FILEDEF);
+}
+
+static upb_filedef *lupb_filedef_checkmutable(lua_State *L, int narg) {
+  const upb_filedef *f = lupb_filedef_check(L, narg);
+  if (upb_filedef_isfrozen(f))
+    luaL_error(L, "not allowed on frozen value");
+  return (upb_filedef*)f;
+}
+
+static int lupb_filedef_new(lua_State *L) {
+  upb_filedef *f = upb_filedef_new(&f);
+  lupb_filedef_pushnewrapper(L, f, &f);
+  return 1;
+}
+
+static int lupb_filedef_def(lua_State *L) {
+  const upb_filedef *f = lupb_filedef_check(L, 1);
+  int index = luaL_checkint(L, 2);
+  const upb_def *def = upb_filedef_def(f, index);
+
+  if (!def) {
+    return luaL_error(L, "index out of range");
+  }
+
+  lupb_def_pushwrapper(L, def, NULL);
+  return 1;
+}
+
+static int lupb_filedef_name(lua_State *L) {
+  const upb_filedef *f = lupb_filedef_check(L, 1);
+  lua_pushstring(L, upb_filedef_name(f));
+  return 1;
+}
+
+static int lupb_filedef_package(lua_State *L) {
+  const upb_filedef *f = lupb_filedef_check(L, 1);
+  lua_pushstring(L, upb_filedef_package(f));
+  return 1;
+}
+
+static int lupb_filedef_len(lua_State *L) {
+  const upb_filedef *f = lupb_filedef_check(L, 1);
+  lua_pushinteger(L, upb_filedef_defcount(f));
+  return 1;
+}
+
+static int lupb_filedef_setname(lua_State *L) {
+  upb_filedef *f = lupb_filedef_checkmutable(L, 1);
+  CHK(upb_filedef_setname(f, lupb_checkname(L, 2), &status));
+  return 0;
+}
+
+static int lupb_filedef_setpackage(lua_State *L) {
+  upb_filedef *f = lupb_filedef_checkmutable(L, 1);
+  CHK(upb_filedef_setpackage(f, lupb_checkname(L, 2), &status));
+  return 0;
+}
+
+static int lupb_filedefiter_next(lua_State *L) {
+  const upb_filedef *f = lupb_filedef_check(L, lua_upvalueindex(1));
+  int type = lua_tointeger(L, lua_upvalueindex(2));
+  size_t i = lua_tointeger(L, lua_upvalueindex(3));
+  size_t n = upb_filedef_defcount(f);
+
+  for (; i < n; i++) {
+    const upb_def *def;
+
+    def = upb_filedef_def(f, i);
+    assert(def);
+
+    if (type == UPB_DEF_ANY || upb_def_type(def) == type) {
+      lua_pushinteger(L, i + 1);
+      lua_replace(L, lua_upvalueindex(3));
+      lupb_def_pushwrapper(L, def, NULL);
+      return 1;
+    }
+  }
+
+  return 0;
+}
+
+static int lupb_filedef_defs(lua_State *L) {
+  lupb_filedef_check(L, 1);
+  luaL_checkinteger(L, 2);   /* Def type.  Could make this optional. */
```

Comment:
It would be easy, I'm still trying to decide whether it's desirable. Maybe the explicitness of file:defs(upb.DEF_ANY) is good? Weighing advice like: https://google.github.io/styleguide/cppguide.html#Default_Arguments

What do you think?